clippy_lints/src/types.rs

```diff
@@ -1210,21 +1211,27 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for Casts {
             let (cast_from, cast_to) = (cx.tables.expr_ty(ex), cx.tables.expr_ty(expr));
             lint_fn_to_numeric_cast(cx, expr, ex, cast_from, cast_to);
             if let ExprKind::Lit(ref lit) = ex.kind {
-                if let LitKind::Int(n, _) = lit.node {
-                    if cast_to.is_floating_point() {
+                if_chain! {
+                    if let LitKind::Int(n, _) = lit.node;
+                    if let Some(src) = snippet_opt(cx, lit.span);
+                    if cast_to.is_floating_point();
+                    then {
                         let from_nbits = 128 - n.leading_zeros();
                         let to_nbits = fp_ty_mantissa_nbits(cast_to);
-                        if from_nbits != 0 && to_nbits != 0 && from_nbits <= to_nbits {
-                            span_lint_and_sugg(
-                                cx,
-                                UNNECESSARY_CAST,
-                                expr.span,
-                                &format!("casting integer literal to `{}` is unnecessary", cast_to),
-                                "try",
-                                format!("{}_{}", n, cast_to),
-                                Applicability::MachineApplicable,
-                            );
-                            return;
+                        if let Some(num_lit) = NumericLiteral::from_lit_kind(&src, &lit.node) {
+                            if from_nbits != 0 && to_nbits != 0 && from_nbits <= to_nbits &&
+                            num_lit.radix != Radix::Hexadecimal {
```

Comment:
I think we also want to _not_ lint this for `0o` and `0b`, so you could add a `is_decimal()` method to the `NumericLiteral` struct.

Reply:
For `0o` and `0b`, the suffixes `_f32` or `_f64` would not be interpreted as part of the literal value such that the cast is indeed unnecessary. Am I correct?
However, it would probably make sense to print the suggestion also in octal or binary format. I will work on that and modify the PR accordingly.

Reply:
Same behavior for `0b` and `0o` [Playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=6b0ecbc9c56b03d00903971b15271a32):

```
warning: casting integer literal to `f32` is unnecessary
 --> src/main.rs:2:13
  |
2 |     let _ = 0b10101 as f32;
  |             ^^^^^^^^^^^^^^ help: try: `21_f32`
  |
  = note: `#[warn(clippy::unnecessary_cast)]` on by default
```

If any other radix, than decimal is used, it is most definitely intended, so suggesting to change it to decimal+`_f32` is not something we want to do. Since binary and octal float literals are not supported, we also want to bail out there.

Reply:
Ok, thanks for the additional pointers. I will include the binary and octal case.

Reply:
Done.